<a href="https://colab.research.google.com/github/pranuk050-pixel/Python_learning-consepts/blob/main/Functions_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
### scope of the variables

### local
### encolsing brackets
### global varaibles


In [ ]:
name="Data Engineering" ### global varaible
def outer_function():
  name="Data science"
  def inner_function():
    nonlocal name
    name="Artificial Intelligence"
  inner_function()
  print(name)
outer_function()

Artificial Intelligence


In [ ]:
def outer():

    x = 10

    def inner():
        nonlocal x
        x = 20

    inner()

    print(x)

outer()

20


In [ ]:
x = 100      # Global variable
def outer():
    y = 200   # Enclosing variable
    def inner():
        global x
        nonlocal y
        x = x + 50
        y = y + 50
        print("Inside Inner")
        print("Global x :", x)
        print("Enclosing y :", y)
    inner()
    print("Inside Outer")
    print("y =", y)
outer()
print("Outside")
print("x =", x)

Inside Inner
Global x : 150
Enclosing y : 250
Inside Outer
y = 250
Outside
x = 150


In [ ]:
count = 0
def company():
    employees = 10
    def hire():
        global count
        nonlocal employees
        count += 1
        employees += 5
        print("Employees:", employees)
        print("Global Count:", count)
    hire()
    print("Final Employees:", employees)
company()
print("Total Companies Processed:", count)

Employees: 15
Global Count: 1
Final Employees: 15
Total Companies Processed: 1


In [ ]:
bank_balance = 10000
def account():
    savings = 5000
    def deposit():
        global bank_balance
        nonlocal savings
        bank_balance += 2000
        savings += 1000
        print("Savings =", savings)
        print("Bank Balance =", bank_balance)
    deposit()
    print("Account Savings =", savings)
account()
print("Global Balance =", bank_balance)

Savings = 6000
Bank Balance = 12000
Account Savings = 6000
Global Balance = 12000


In [ ]:
x = 10
def outer():
    y = 20
    def middle():
        z = 30
        def inner():
            global x
            nonlocal y
            nonlocal z
            x += 1
            y += 2
            z += 3
            print("Inner")
            print(x, y, z)
        inner()
        print("Middle:", z)
    middle()
    print("Outer:", y)
outer()
print("Global:", x)

Inner
11 22 33
Middle: 33
Outer: 22
Global: 11


In [ ]:
### global : varaibles declared outside all functions : globa
### nonlocal: varaible declared in the nearest to the enclosing functions, nonlocal
### LEG : local, enclosing brackets, global

# Machine Learning Notes with Examples

# 1. Derived Columns

## Definition
Derived columns are new columns created from existing columns using calculations, formulas, or transformations.

## Why?
- Improve model performance
- Capture hidden patterns
- Create meaningful business features

## Examples

### Customer Age Group
```python
Age = 28

if Age < 18:
    Age_Group = "Child"
elif Age < 60:
    Age_Group = "Adult"
else:
    Age_Group = "Senior"
```

### Profit
```
Profit = Sales - Cost
```

### Total Amount
```
Total = Quantity × Price
```

### BMI
```
BMI = Weight / Height²
```

---

# 2. Feature Engineering

## Definition

Feature Engineering is the process of creating, modifying, or selecting features that help machine learning models perform better.

## Types

- Derived Features
- Feature Encoding
- Date Features
- Polynomial Features
- Interaction Features
- Scaling
- Binning

## Example

Original Data

| DOB | Salary |
|------|---------|
|1998-03-12|45000|

Engineered Features

|Age|Salary|Salary_Level|
|---|------|------------|
|27|45000|Medium|

Python

```python
df["Age"] = 2025 - df["BirthYear"]
df["Salary_Level"] = df["Salary"] > 50000
```

---

# 3. Feature Weightage

## Definition

Feature weightage indicates how important each feature is in predicting the target.

## Example

House Price Prediction

|Feature|Importance|
|--------|----------|
|Area|0.48|
|Bedrooms|0.22|
|Location|0.18|
|Parking|0.08|
|Age|0.04|

Higher importance means greater influence.

---

# 4. Clustering

## Definition

Clustering groups similar data points without knowing the labels.

It is an Unsupervised Learning technique.

## Algorithms

- K-Means
- DBSCAN
- Hierarchical Clustering
- Mean Shift

## Example

Customer Data

|Customer|Annual Income|Spending|
|----------|------------|---------|
|A|40k|80|
|B|42k|82|
|C|100k|20|

Output

Cluster 1
- A
- B

Cluster 2
- C

---

# 5. Segmentation

## Definition

Segmentation divides data into meaningful groups.

## Examples

Customer Segmentation

- Premium Customers
- Regular Customers
- New Customers

Marketing Segmentation

- Teenagers
- Adults
- Seniors

---

# 6. Dimensionality Reduction

## Definition

Reducing the number of features while preserving maximum information.

## Benefits

- Faster training
- Less memory
- Better visualization
- Removes noise

## Methods

- PCA
- LDA
- t-SNE
- UMAP

Example

Original Features

```
100 Columns
```

Reduced

```
15 Columns
```

---

# 7. Feature Reduction

## Definition

Removing unnecessary features.

## Why?

- Remove duplicate columns
- Remove constant columns
- Remove highly correlated columns
- Reduce overfitting

Example

Original

```
Age
DOB
BirthYear
```

Keep only

```
Age
```

---

# 8. Continuous Variables

## Definition

Variables that can take any value.

## Examples

- Height
- Weight
- Temperature
- Salary
- Rainfall

Example

```
45.78

82.33

168.25
```

---

# 9. Categorical Variables

## Definition

Variables representing categories.

Examples

- Male
- Female

- Red
- Green

- Yes
- No

---

# 10. Discrete Variables

## Definition

Variables with countable values.

Examples

- Number of Students
- Number of Cars
- Number of Employees

Possible Values

```
1
2
3
10
15
```

Not

```
2.56
```

---

# 11. Ordinal Variables

## Definition

Categories having an order.

Examples

Education

```
Primary
Secondary
Graduate
Masters
PhD
```

Customer Rating

```
Poor
Average
Good
Excellent
```

---

# 12. Nominal Variables

## Definition

Categories without order.

Examples

- Country
- City
- Blood Group
- Department

Example

```
HR
IT
Finance
Marketing
```

---

# 13. Date Variables

## Definition

Variables containing calendar dates.

Examples

```
2025-01-10

2026-07-28

2024-12-31
```

Common Features Extracted

- Year
- Month
- Day
- Week
- Quarter
- Weekend

Python

```python
df["Year"] = df["Date"].dt.year
df["Month"] = df["Date"].dt.month
```

---

# 14. Time Series Variables

## Definition

Data recorded over time.

Examples

- Stock Price
- Weather
- Sales
- Electricity Usage

Example

|Date|Sales|
|----|-----|
|Jan|100|
|Feb|120|
|Mar|150|

---

# 15. Date and Timestamp

## Date

```
2026-07-28
```

Contains

- Year
- Month
- Day

---

## Timestamp

```
2026-07-28 14:35:18
```

Contains

- Year
- Month
- Day
- Hour
- Minute
- Second

---

# 16. Data Collection

## Definition

Collecting data from different sources.

## Common Sources

### CSV

```
employees.csv
```

### TXT

```
employees.txt
```

### Database

```
MySQL

SQL Server

PostgreSQL

Oracle
```

### JSON

```json
{
"name":"John",
"age":30
}
```

### APIs

```
REST API

GraphQL
```

### Cloud Storage

- Azure Blob Storage
- AWS S3
- Google Cloud Storage

---

# 17. Preprocessing

## Definition

Cleaning and preparing data before training.

### Cleaning

Remove

- Missing Values
- Duplicate Records
- Invalid Values
- Outliers

Example

Before

```
100
NULL
150
150
```

After

```
100
120
150
```

---

### Extraction

Extract useful information.

Example

```
Email

john@gmail.com

Extract

gmail.com
```

---

### Transformation

Convert data into suitable format.

Examples

- Scaling
- Encoding
- Normalization

Example

```
Male → 0

Female → 1
```

---

### Loading

Store processed data.

Destination

- SQL Database
- Data Warehouse
- Azure Data Lake
- Snowflake

---

# 18. EDA (Exploratory Data Analysis)

## Definition

Understanding data before building models.

## Activities

- Summary Statistics
- Missing Values
- Correlation
- Distribution
- Visualization

Charts

- Histogram
- Box Plot
- Scatter Plot
- Heatmap
- Pair Plot
- Bar Chart

Questions Answered

- Is data clean?
- Are there outliers?
- Which features are correlated?

---

# 19. Model Building

## Definition

Training a machine learning algorithm.

Steps

1. Split Data
2. Select Algorithm
3. Train Model
4. Predict

Algorithms

Classification

- Logistic Regression
- Decision Tree
- Random Forest

Regression

- Linear Regression
- Ridge
- Lasso

Clustering

- K-Means
- DBSCAN

---

# 20. Model Validation

## Definition

Checking model performance on unseen data.

Methods

- Train-Test Split
- Cross Validation
- Stratified Split

Metrics

Classification

- Accuracy
- Precision
- Recall
- F1 Score
- ROC-AUC

Regression

- MAE
- RMSE
- MSE
- R² Score

---

# 21. Model Fine-Tuning

## Definition

Improving model performance by optimizing hyperparameters.

Methods

- Grid Search
- Random Search
- Bayesian Optimization

Example

Random Forest

```
n_estimators

max_depth

min_samples_split
```

---

# 22. Model Deployment

## Definition

Making a trained model available for real-world use.

Deployment Options

- Flask API
- FastAPI
- Azure ML
- Docker
- Kubernetes
- AWS SageMaker
- Azure App Service

Workflow

```
Train Model

↓

Save Model

↓

Create API

↓

Deploy

↓

Users Send Requests

↓

Predictions Returned
```

---

# 23. Model Monitoring

## Definition

Continuously tracking model performance after deployment.

Monitor

- Accuracy
- Latency
- Drift
- Errors
- Resource Usage

Example

```
Training Accuracy

95%

Production Accuracy

80%

↓

Model Drift Detected

↓

Retrain Model
```

---

# 24. Population

## Definition

The complete set of all observations that you want to study.

Examples

- All customers of Amazon
- All students in a university
- All employees in a company
- Every transaction in a bank

Example

Population = 50,000 customers

---

# 25. Sample

## Definition

A subset selected from the population for analysis.

Example

Population

```
50,000 Customers
```

Sample

```
5,000 Customers
```

Why Sample?

- Lower cost
- Faster analysis
- Easier computation

---

# Population vs Sample

|Feature|Population|Sample|
|--------|----------|------|
|Definition|Entire dataset|Subset of dataset|
|Size|Large|Smaller|
|Cost|High|Low|
|Time|More|Less|
|Purpose|Complete study|Estimate population characteristics|

Example

```
Population

All 1,00,000 bank customers

↓

Randomly select

↓

10,000 customers

↓

Sample
```

---

# Complete Machine Learning Workflow

```
Problem Statement
        │
        ▼
Data Collection
        │
        ▼
Data Cleaning
        │
        ▼
Feature Engineering
        │
        ▼
Feature Selection / Reduction
        │
        ▼
EDA
        │
        ▼
Train-Test Split
        │
        ▼
Model Building
        │
        ▼
Model Validation
        │
        ▼
Hyperparameter Fine-Tuning
        │
        ▼
Model Deployment
        │
        ▼
Model Monitoring
        │
        ▼
Retraining (if required)
```